File: cpp/src/frame.cpp
#include "arnio/frame.h"

#include <stdexcept>

namespace arnio {

Frame::Frame(size_t row_count) : row_count_(row_count), row_count_known_(true) {}

Frame::Frame(std::vector<Column> columns) : columns_(std::move(columns)) {
    if (!columns_.empty()) {
        row_count_ = columns_[0].size();
        for (const auto& col : columns_) {
            validate_column_size(col);
        }
    }
    row_count_known_ = true;
    rebuild_index();
}

Frame::Frame(size_t row_count, std::vector<Column> columns)
    : columns_(std::move(columns)), row_count_(row_count), row_count_known_(true) {
    for (const auto& col : columns_) {
        validate_column_size(col);
    }
    rebuild_index();
}

std::pair<size_t, size_t> Frame::shape() const { return {num_rows(), num_cols()}; }

size_t Frame::num_rows() const { return row_count_; }

size_t Frame::num_cols() const { return columns_.size(); }

std::vector<std::string> Frame::column_names() const {
    std::vector<std::string> names;
    names.reserve(columns_.size());
    for (const auto& col : columns_) {
        names.push_back(col.name());
    }
    return names;
}

std::unordered_map<std::string, std::string> Frame::dtypes() const {
    std::unordered_map<std::string, std::string> result;
    for (const auto& col : columns_) {
        result[col.name()] = dtype_to_string(col.dtype());
    }
    return result;
}

size_t Frame::memory_usage() const {
    size_t usage = sizeof(Frame);
    for (const auto& col : columns_) {
        usage += col.memory_usage();
    }
    return usage;
}

const Column& Frame::column(size_t idx) const {
    if (idx >= columns_.size()) {
        throw std::out_of_range("Column index out of range");
    }
    return columns_[idx];
}

const Column& Frame::column(const std::string& name) const {
    auto it = name_index_.find(name);
    if (it == name_index_.end()) {
        throw std::out_of_range("Column not found: " + name);
    }
    return columns_[it->second];
}

bool Frame::has_column(const std::string& name) const {
    return name_index_.find(name) != name_index_.end();
}

size_t Frame::column_index(const std::string& name) const {
    auto it = name_index_.find(name);
    if (it == name_index_.end()) {
        throw std::out_of_range("Column not found: " + name);
    }
    return it->second;
}

void Frame::add_column(Column col) {
    if (!row_count_known_) {
        row_count_ = col.size();
        row_count_known_ = true;
    } else {
        validate_column_size(col);
    }
    name_index_[col.name()] = columns_.size();
    columns_.push_back(std::move(col));
}

const std::vector<Column>& Frame::columns() const { return columns_; }

Frame Frame::clone() const {
    std::vector<Column> cloned;
    cloned.reserve(columns_.size());
    for (const auto& col : columns_) {
        cloned.push_back(col.clone());
    }
    return Frame(row_count_, std::move(cloned));
}

void Frame::validate_column_size(const Column& col) const {
    if (col.size() != row_count_) {
        throw std::invalid_argument("Column '" + col.name() + "' has row count " +
                                    std::to_string(col.size()) + "; expected " +
                                    std::to_string(row_count_));
    }
}

Frame Frame::select_columns(const std::vector<std::string>& columns) const {
    std::vector<Column> selected;
    selected.reserve(columns.size());

    for (const auto& name : columns) {
        selected.push_back(column(name).clone());
    }

    return Frame(row_count_, std::move(selected));
}

void Frame::rebuild_index() {
    name_index_.clear();
    for (size_t i = 0; i < columns_.size(); ++i) {
        name_index_[columns_[i].name()] = i;
    }
}

}  // namespace arnio

File: arnio/frame.py
"""
arnio.frame
ArFrame — the core data container wrapping the C++ Frame.
"""

from __future__ import annotations

from ._core import _Frame

#: Dtype strings recognised by ArFrame.select_dtypes().
_VALID_DTYPES: frozenset[str] = frozenset(
    {"int64", "float64", "string", "bool", "null"}
)


class ArFrame:
    """Lightweight columnar data container backed by C++."""

    __slots__ = ("_frame", "_attrs")

    def __init__(self, cpp_frame: _Frame, attrs: dict | None = None) -> None:
        self._frame = cpp_frame
        self._attrs: dict = attrs if attrs is not None else {}

    # --- Properties ---

    @property
    def shape(self) -> tuple[int, int]:
        """Row and column count.

        Returns
        -------
        tuple[int, int]
            (number_of_rows, number_of_columns)
        """
        return self._frame.shape()

    @property
    def columns(self) -> list[str]:
        """Column names.

        Returns
        -------
        list[str]
            List of column names in order.
        """
        return self._frame.column_names()

    @property
    def dtypes(self) -> dict[str, str]:
        """Column name → inferred type.

        Returns
        -------
        dict[str, str]
            Mapping of column names to their data types.
        """
        return self._frame.dtypes()

    @property
    def is_empty(self) -> bool:
        """Check if frame has zero rows.

        Returns
        -------
        bool
            True if frame contains no rows, False otherwise.

        Examples
        --------
        >>> frame = ar.read_csv("data.csv")
        >>> if frame.is_empty:
        ...     print("No data to process")
        False
        """
        return len(self) == 0

    # --- Methods ---

    def memory_usage(self) -> int:
        """Total bytes consumed in memory.

        Returns
        -------
        int
            Memory usage in bytes.
        """
        return self._frame.memory_usage()

    def head(self, n: int = 5) -> ArFrame:
        """Return the first n rows as an ArFrame.

        Parameters
        ----------
        n : int, optional
            Number of rows to return. Defaults to 5.

        Returns
        -------
        ArFrame
            New ArFrame containing the first n rows.
        """
        if isinstance(n, bool) or not isinstance(n, int) or n < 0:
            raise ValueError(f"`n` must be a non-negative integer, got {n!r}")

        from .convert import from_pandas, to_pandas

        df = to_pandas(self)

        return from_pandas(df.head(n))

    def tail(self, n: int = 5) -> ArFrame:
        """Return the last n rows as an ArFrame.

        Parameters
        ----------
        n : int, optional
            Number of rows to return. Defaults to 5.

        Returns
        -------
        ArFrame
            New ArFrame containing the last n rows.
        """
        if isinstance(n, bool) or not isinstance(n, int) or n < 0:
            raise ValueError(f"`n` must be a non-negative integer, got {n!r}")

        from .convert import from_pandas, to_pandas

        df = to_pandas(self)

        return from_pandas(df.tail(n))

    def select_columns(self, columns: list[str]) -> ArFrame:
        """Return a new ArFrame with only the selected columns.

        Parameters
        ----------
        columns : list[str]
            List of column names to select.

        Returns
        -------
        ArFrame
            New ArFrame containing only the selected columns.

        Raises
        ------
        TypeError
            If columns is not a valid sequence of strings.
        ValueError
            If the selection is empty, contains duplicates,
            or includes unknown columns.
        """
        if isinstance(columns, str):
            raise TypeError("columns must be a sequence of column names, not a string.")

        if not isinstance(columns, (list, tuple)):
            raise TypeError("columns must be a list or tuple of column names.")

        if not columns:
            raise ValueError("Column selection cannot be empty.")

        if any(not isinstance(col, str) for col in columns):
            raise TypeError("All column names must be strings.")

        if len(columns) != len(set(columns)):
            raise ValueError("Duplicate column names are not allowed.")

        missing = [col for col in columns if col not in self.columns]

        if missing:
            raise ValueError(f"Unknown columns: {missing}")

        return ArFrame(self._frame.select_columns(columns))

    def select_dtypes(
        self,
        include: str | list[str] | tuple[str, ...] | None = None,
        exclude: str | list[str] | tuple[str, ...] | None = None,
    ) -> ArFrame:
        """Return a new ArFrame containing only columns whose dtype matches the filter.

        At least one of *include* or *exclude* must be provided.

        Parameters
        ----------
        include : str, list[str], or tuple[str, ...], optional
            One or more dtype strings to keep.
            Accepted values: ``"int64"``, ``"float64"``, ``"string"``,
            ``"bool"``, ``"null"``.
        exclude : str, list[str], or tuple[str, ...], optional
            One or more dtype strings to drop. Applied after *include*.

        Returns
        -------
        ArFrame
            New ArFrame containing only the matched columns, in original
            column order.

        Raises
        ------
        ValueError
            If neither *include* nor *exclude* is provided, if *include*
            and *exclude* overlap, if an unrecognised dtype string is
            passed, or if no columns match the filter.
        TypeError
            If *include* or *exclude* is not a string, list, or tuple of
            strings.

        Examples
        --------
        >>> frame = ar.read_csv("data.csv")
        >>> numeric = frame.select_dtypes(include=["int64", "float64"])
        >>> without_strings = frame.select_dtypes(exclude="string")
        """
        if include is None and exclude is None:
            raise ValueError(
                "select_dtypes() requires at least one of 'include' or 'exclude'."
            )

        def _parse(
            arg: str | list[str] | tuple[str, ...] | None,
            name: str,
        ) -> frozenset[str] | None:
            if arg is None:
                return None
            if isinstance(arg, str):
                values = [arg]
            elif isinstance(arg, (list, tuple)):
                values = list(arg)
                non_strings = [v for v in values if not isinstance(v, str)]
                if non_strings:
                    raise TypeError(
                        f"'{name}' must contain only strings, "
                        f"got {[type(v).__name__ for v in non_strings]}."
                    )
            else:
                raise TypeError(
                    f"'{name}' must be a string, list, or tuple of strings, "
                    f"got {type(arg).__name__!r}."
                )
            unknown = [v for v in values if v not in _VALID_DTYPES]
            if unknown:
                raise ValueError(
                    f"Unrecognised dtype(s) in '{name}': {unknown}. "
                    f"Valid dtypes are: {sorted(_VALID_DTYPES)}."
                )
            return frozenset(values)

        include_set = _parse(include, "include")
        exclude_set = _parse(exclude, "exclude")

        if include_set is not None and exclude_set is not None:
            overlap = include_set & exclude_set
            if overlap:
                raise ValueError(
                    f"'include' and 'exclude' overlap: {sorted(overlap)}. "
                    "A dtype cannot be both included and excluded."
                )

        col_dtypes = self.dtypes
        matched: list[str] = []
        for col in self.columns:  # iterate columns to preserve original order
            dtype = col_dtypes[col]
            if include_set is not None and dtype not in include_set:
                continue
            if exclude_set is not None and dtype in exclude_set:
                continue
            matched.append(col)

        if not matched:
            raise ValueError(
                "No columns match the dtype selection. " f"Frame dtypes: {col_dtypes}."
            )

        return self.select_columns(matched)

    def _truncate_column_names(self, max_length=20):
        return [
            col[:max_length] + "..." if len(col) > max_length else col
            for col in self.columns
        ]

    # --- Dunder methods ---

    def __len__(self) -> int:
        """Return the number of rows."""
        return self._frame.num_rows()

    def __repr__(self) -> str:
        """Return a string representation of the ArFrame."""
        rows, cols = self.shape
        return f"ArFrame({rows} rows × {cols} cols)"

    def __str__(self) -> str:
        """Return a detailed string summary of the ArFrame."""
        lines = [f"ArFrame: {self.shape[0]} rows × {self.shape[1]} columns"]
        lines.append(f"Columns: {self._truncate_column_names()}")
        lines.append(f"DTypes:  {self.dtypes}")
        lines.append(f"Memory:  {self.memory_usage()} bytes")
        return "\n".join(lines)

    def __contains__(self, item: object) -> bool:
        return isinstance(item, str) and item in self.columns

    def __getitem__(self, key: str) -> list:
        """Return column data as a list.

        Parameters
        ----------
        key : str
            Column name to access.

        Returns
        -------
        list
            Column values as a Python list.

        Raises
        ------
        TypeError
            If key is not a string.
        KeyError
            If the column does not exist.

        Examples
        --------
        >>> frame = ar.read_csv("data.csv")
        >>> frame["name"]
        ['Alice', 'Bob', 'Charlie']
        """
        if not isinstance(key, str):
            raise TypeError(f"column key must be a string, got {type(key).__name__!r}")

        if key not in self.columns:
            raise KeyError(
                f"Column {key!r} not found. Available columns: {self.columns}"
            )

        col_index = self.columns.index(key)
        return [self._frame.column_by_index(col_index).at(i) for i in range(len(self))]

    def preview(self, n: int = 5) -> str:
        """Return a lightweight string preview of the first ``n`` rows.

        Reads only the first ``n`` rows directly from the C++ frame without
        triggering a full pandas conversion, making it safe to call on very
        large frames from the CLI or a notebook.

        Parameters
        ----------
        n : int, optional
            Number of rows to preview. Must be a positive integer.
            Defaults to 5.

        Returns
        -------
        str
            A formatted string table showing the first ``n`` rows.

        Raises
        ------
        ValueError
            If ``n`` is not a positive integer.

        Examples
        --------
        >>> frame = ar.read_csv("data.csv")
        >>> print(frame.preview())       # first 5 rows
        >>> print(frame.preview(n=10))   # first 10 rows
        """
        if isinstance(n, bool) or not isinstance(n, int) or n < 1:
            raise ValueError(f"`n` must be a positive integer, got {n!r}")

        num_rows, num_cols = self.shape

        if num_rows == 0:
            return "ArFrame preview: (empty frame)"

        actual_n = min(n, num_rows)

        # Pull only the first `actual_n` values per column — no full conversion
        col_names = self.columns
        col_data = [
            [self._frame.column_by_index(i).at(r) for r in range(actual_n)]
            for i in range(num_cols)
        ]

        # Calculate column widths for alignment
        col_widths = [
            max(
                len(col_names[i]),
                max((len(str(col_data[i][r])) for r in range(actual_n)), default=0),
            )
            for i in range(num_cols)
        ]

        # Build header and separator
        header = "  ".join(col_names[i].ljust(col_widths[i]) for i in range(num_cols))
        separator = "  ".join("-" * col_widths[i] for i in range(num_cols))

        # Build rows
        rows = [
            "  ".join(str(col_data[i][r]).ljust(col_widths[i]) for i in range(num_cols))
            for r in range(actual_n)
        ]

        label = f"ArFrame preview (showing {actual_n} of {num_rows} rows):"
        return "\n".join([label, header, separator] + rows)
